packageurl.go

```diff
@@ -331,6 +354,43 @@ func typeAdjustName(purlType, name string) string {
 	return name
 }
 
+// validQualifierKey validates a qualifierKey against our QualifierKeyPattern.
 func validQualifierKey(key string) bool {
 	return QualifierKeyPattern.MatchString(key)
 }
+
+// validCustomRules evaluates additional rules for each package url type, as specified in the package-url specification.
+// On success, it returns nil. On failure, a descriptive error will be returned.
+func validCustomRules(purlType, name, ns, version string, qualifiers Qualifiers) error {
```

Comment:
Would it make sense to add these to the type spec document?

Reply:
The rules itself are already mentioned in the type spec document, they are just not written as code or in a clear manner
For example for the type [`conan`](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#conan)   the specification says: The qualifier channel must be not empty if namespace is present.